### PR TITLE
Define API routes for the Quiz application

### PR DESCRIPTION
**Pull request by parth-1372**
**Overview**
This Pull Request defines the API routes needed for the Quiz application. It outlines each endpoint, the HTTP methods used, and their purposes.

**Changes Made**
- Added API routes for user management:
  - **GET** `/api/users`: Fetch all users with pagination (query params: `page`, `limit`).
  - **POST** `/api/users`: Create a new user (request body: `name`, `email`).
  - **GET** `/api/users/:id`: Fetch a user by ID (path param: `id`).
  - **PUT** `/api/users/:id`: Update a user by ID (path param: `id`, request body: `name`, `email`).
  - **DELETE** `/api/users/:id`: Delete a user by ID (path param: `id`).

- Updated authentication routes:
  - **POST** `/api/auth/register`: Register a new user (request body: `name`, `email`, `password`).
  - **POST** `/api/auth/login`: Authenticate a user (request body: `email`, `password`).
  - **POST** `/api/auth/logout`: Log out a user (path param: `id`).

Testing
Postman API Testing: All API routes were tested using Postman to ensure they function as expected. Below are the details of the tests conducted:
[Routes-Testing.pdf](https://github.com/user-attachments/files/17221106/Routes-Testing.pdf)

User Management Routes
**GET /api/users**
-Fetches a list of users with pagination.
-Query Parameters: page (default: 1), limit (default: 10).
-Test Result: Successfully returned the expected user list with correct pagination.

**POST /api/users**
-Description: Creates a new user.
-Request Body:
json
{
  "name": "XYZ",
  "email": "xyz@example.com"
}
Test Result: User created successfully with a confirmation message.

**GET /api/users/123**
-Description: Fetches a user by their ID.
-Path Parameter: id (e.g., 123).
-Test Result: Successfully returned user data for the specified ID.

**PUT /api/users/**
-Description: Updates a user's information.
-Path Parameter: id (e.g., 1).
-Request Body:
json
Copy code
{
  "name": "XYZ Updated",
  "email": "xyz.updated@example.com"
}
-Test Result: User updated successfully with confirmation.

**DELETE /api/users/1**
-Description: Deletes a user by their ID.
-Path Parameter: id (e.g., 1).
-Test Result: User deleted successfully with a confirmation message.

**Authentication Routes**

**POST /api/auth/register**
-Description: Registers a new user.
-Request Body:
{
  "name": "XYZ",
  "email": "xyz@example.com",
  "password": "*********"
}
-Test Result: User registered successfully with a confirmation message.

**POST /api/auth/login**
-Description: Authenticates a user.
-Request Body:
{
  "email": "xyz@example.com",
  "password": "*********"
}
-Test Result: User logged in successfully with a token.

**POST /api/auth/logout**
-Description: Logs out a user.
-Path Parameter: id (e.g., 1).
-Test Result: User logged out successfully with a confirmation message.

### CODE REVIEW

File: src/server/Routes/Auth.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        // Example  bcrypt.hash(password, 10);
+        // Here, you can add the logic to store the user in the database
+        
+        // Simulating a successful registration
+        res.json({ message: "User Registered Successfully", email });
+    } catch (error) {
+        // Simple error response
+        res.status(500).json({ message: "Registration failed. Please try again." });
+    }
+});
```
Comment (shubhansu-kr) [inline src/server/Routes/Auth.js:19]: Please follow the defined structure for creating REST API in the documentations.

Comment (shubhansu-kr) [reply src/server/Routes/Auth.js:19]: Refer to this PR for clarity #15 against this issue #13

**Comment by shubhansu-kr** [comment]
@parth-1372 , Thanks for your contribution !

The scope the issue #10 was to come up with a plan for routes, get it reviewed and then implement it once approved. I can see you have implemented the changes directly. 

Please provide the list of routes which we need to implement like mentioned in the issue description.

File: src/server/index.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 const express = require('express');
 const cors = require('cors');
 const dotenv = require('dotenv');
-
+const {Router }=  require("express");
```
Comment (shubhansu-kr) [inline src/server/index.js:5]: Nit : Use lint before PR. Import statement should be clubbed together. Move it below db import

File: src/server/index.js [hunk 3]
```diff
@@ -11,14 +11,22 @@ dotenv.config();
...
+const AuthRoute = require("./Routes/Auth");
 app.use(cors());
 app.use(express.json());
 
 app.get("/", (req, res)=>{
     res.send("Quikz: Hello World!");
 });
+app.use(express.json()); // to parse application/json
+
+app.use('/api/users', Userroute);
```
Comment (shubhansu-kr) [inline src/server/index.js:27]: This should have been in different file dedicated for routing. We cannot clutter everything in index.js file. Difficult to maintain and scale.

File: src/server/Routes/Auth.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+const express = require('express');
+const router = express.Router();
+
+// Route for registering a user
+router.post("/register", async (req, res) => {
+    const { name, email, password } = req.body;
```
Comment (shubhansu-kr) [inline src/server/Routes/Auth.js:6]: What if req does not have the name, email and password ? Shouldn't we have checks here ?

File: src/server/Routes/Auth.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    try {
+        // Hash the password for security
+        // Example  bcrypt.hash(password, 10);
```
Comment (shubhansu-kr) [inline src/server/Routes/Auth.js:10]: Please implement this, If it's AI generated give it a look before PR

File: src/server/Routes/Auth.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        // Here, you can add the logic to store the user in the database
+        
+        // Simulating a successful registration
+        res.json({ message: "User Registered Successfully", email });
```
Comment (shubhansu-kr) [inline src/server/Routes/Auth.js:14]: Add status code

File: src/server/Routes/Auth.js (hunk 4 above)
Comment (shubhansu-kr) [reply src/server/Routes/Auth.js:6]: Apply whereever applicable throughout PR.

**Review by shubhansu-kr — CHANGES_REQUESTED** [review]
Need a lot of improvements, mentioned few in the comments. 

First and foremost, adhere to repository structure.

**Comment by shubhansu-kr** [comment]
What @mvp-2003 , why would you merge it ?

**Comment by parth-1372** [comment]
Thank you for the clarification, @shubhansu-kr.
I have a question regarding the issue (#13) mentioned. Should I only define the routes for the requested repository, including the status codes, or do I need to implement complete logic with controllers, models, middleware, etc.?

Regarding the hashing logic, I researched popular hashing algorithms(And Take Help Of AI  But don't copy) and chose an approach that best fits our needs. However, I understand that writing detailed logic may not be a part of this issue, so i add comment for Other Developer to use in future.

**Comment by shubhansu-kr** [comment]
We dont want implementation before we agree on the routing plan. Please add a doc to the isssue #10 , for what all routes you want to create to serve all the use case of our application. 

I understand that we have not outlined the use case but that is rest with you to define what should be the standard uses of a quiz application. Like you added, User and Auth, you can come up with further uses like Quiz types, Quiz model, participant, host, etc..

**Comment by shubhansu-kr** [comment]
Simply think from the product perspective what you need and create a table of all the routes you would need as mentioned in issue Description.

**Comment by shubhansu-kr** [comment]
> Should I only define the routes for the requested repository, including the status codes, or do I need to implement complete logic with controllers, models, middleware, etc.?

We will create sub tasks for each routes implementation. Need not code anything rn

**Pull request merged**